File: December - 30/KishoreMuruganantham_burn_binary_tree.swift
import Foundation
var lines = [String]()
while let line = readLine() { lines.append(line) }
if !lines.isEmpty {
    var vals = [String]()
    var target = 0
    
    for line in lines {
        if line.contains("root") {
            var s = line
            if let r = s.range(of: "=") { s = String(s[r.upperBound...]) }
            s = s.replacingOccurrences(of: "[", with: "")
            s = s.replacingOccurrences(of: "]", with: "")
            s = s.replacingOccurrences(of: " ", with: "")
            vals = s.split(separator: ",").map { String($0) }
        } else if line.contains("Target") {
            var s = line
            if let r = s.range(of: "=") { s = String(s[r.upperBound...]) }
            target = Int(s.trimmingCharacters(in: .whitespaces)) ?? 0
        }
    }
    
    var adj = [Int: [Int]]()
    
    if !vals.isEmpty && vals[0] != "null" {
        let rootVal = Int(vals[0])!
        var q = [rootVal]
        var head = 0
        var idx = 1
        
        while head < q.count && idx < vals.count {
            let curr = q[head]
            head += 1
            
            func addEdge(_ u: Int, _ v: Int) {
                adj[u, default: []].append(v)
                adj[v, default: []].append(u)
            }
            
            if idx < vals.count {
                let val = vals[idx]
                idx += 1
                if val != "null" {
                    let v = Int(val)!
                    addEdge(curr, v)
                    q.append(v)
                }
            }
            
            if idx < vals.count {
                let val = vals[idx]
                idx += 1
                if val != "null" {
                    let v = Int(val)!
                    addEdge(curr, v)
                    q.append(v)
                }
            }
        }
    }
    
    var visited = Set<Int>()
    var qBFS = [target]
    visited.insert(target)
    
    while !qBFS.isEmpty {
        let level = qBFS
        print(level.map{ String($0) }.joined(separator: ", "))
        
        var nextQ = [Int]()
        for u in level {
            if let nbs = adj[u] {
                for nb in nbs {
                    if !visited.contains(nb) {
                        visited.insert(nb)
                        nextQ.append(nb)
                    }
                }
            }
        }
        qBFS = nextQ
    }
}

File: December - 30/KishoreMuruganantham_burn_binary_tree.R
input <- file("stdin", "r")
lines <- readLines(input)

root_idx <- grep("root", lines)
target_idx <- grep("Target", lines)

if (length(root_idx) > 0) {
  root_line <- lines[root_idx[1]]
  line_r <- sub(".*=\\s*", "", root_line)
  line_r <- gsub("\\[|\\]", "", line_r)
  vals <- unlist(strsplit(line_r, ","))
  vals <- gsub("\\s+", "", vals)
} else {
  vals <- character(0)
}

if (length(target_idx) > 0) {
  t_line <- lines[target_idx[1]]
  target <- as.integer(sub(".*=\\s*", "", t_line))
} else {
  target <- 0
}

adj <- new.env(hash = TRUE)

if (length(vals) > 0 && vals[1] != "null") {
  root_val <- as.integer(vals[1])
  q <- list(root_val)
  head <- 1
  idx <- 2
  
  while (head <= length(q) && idx <= length(vals)) {
    curr <- q[[head]]
    head <- head + 1
    
    add_edge <- function(u, v) {
      if (is.null(adj[[as.character(u)]])) adj[[as.character(u)]] <- c()
      adj[[as.character(u)]] <- c(adj[[as.character(u)]], v)
      if (is.null(adj[[as.character(v)]])) adj[[as.character(v)]] <- c()
      adj[[as.character(v)]] <- c(adj[[as.character(v)]], u)
    }
    
    if (idx <= length(vals)) {
      val <- vals[idx]
      idx <- idx + 1
      if (val != "null") {
        v <- as.integer(val)
        add_edge(curr, v)
        q <- c(q, list(v))
      }
    }
    
    if (idx <= length(vals)) {
      val <- vals[idx]
      idx <- idx + 1
      if (val != "null") {
        v <- as.integer(val)
        add_edge(curr, v)
        q <- c(q, list(v))
      }
    }
  }
}

visited <- new.env(hash = TRUE)
q <- list(target)
visited[[as.character(target)]] <- TRUE

while (length(q) > 0) {
  level <- integer(0)
  sz <- length(q)
  
  new_q <- list()
  
  for (item in q) {
    u <- item
    level <- c(level, u)
    
    nbs <- adj[[as.character(u)]]
    for (nb in nbs) {
      if (is.null(visited[[as.character(nb)]])) {
        visited[[as.character(nb)]] <- TRUE
        new_q <- c(new_q, list(nb))
      }
    }
  }
  cat(paste(level, collapse = ", "), "\n", sep = "")
  q <- new_q
}
